aten/src/TH/THTensorApply.h

```diff
@@ -235,4 +245,238 @@
+    int TH_TENSOR_APPLY_hasFinished = 0; \
+    int64_t TH_TENSOR_dim_index = 0; \
+    __TH_TENSOR_APPLYX_PREAMBLE(TYPE2, TENSOR2, -1, 1) \
+    __TH_TENSOR_APPLYX_PREAMBLE(TYPE1, TENSOR1, -1, 1) \
+    PRAGMA(omp parallel if (SIZE > TH_OMP_OVERHEAD_THRESHOLD_OMP) firstprivate(TENSOR2##_sizes, TENSOR2##_strides, TENSOR2##_dim, TENSOR2##_stride, TENSOR2##_size, TENSOR2##_i, TENSOR1##_sizes, TENSOR1##_strides, TENSOR1##_dim, TENSOR1##_stride, TENSOR1##_size, TENSOR1##_i)) \
+    {                                                                                                    \
+      size_t num_threads = omp_get_num_threads();                                                        \
+      size_t tid = omp_get_thread_num();                                                                 \
+      ptrdiff_t line_index_offset = tid * (SIZE/num_threads);                                            \
+      ptrdiff_t line_index_end = (tid == num_threads - 1) ? SIZE:(line_index_offset + SIZE/num_threads); \
+      ptrdiff_t line_seg_len = line_index_end - line_index_offset;                                       \
+      /*divided segment maybe does not cover the size of the last contiguous dimension*/                 \
+      /*calculate each coord of the preamble tensor*/                                                    \
+      __TH_TENSOR_APPLYX_CAL_OFFSET(TENSOR2);                                                            \
+      __TH_TENSOR_APPLYX_CAL_OFFSET(TENSOR1);                                                            \
+      TYPE2 *TENSOR2##_data = tp + TENSOR2##_offset;                                                     \
+      TYPE1 *TENSOR1##_data = rp + TENSOR1##_offset;                                                     \
+      ptrdiff_t count = 0;                                                                               \
+      ptrdiff_t TENSOR2##_start =  TENSOR2##_counter_tmp[TENSOR2##_dim-1];                               \
+      ptrdiff_t TENSOR1##_start =  TENSOR1##_counter_tmp[TENSOR1##_dim-1];                               \
+      while (count < line_seg_len) {                                                                     \
+        for(TENSOR2##_i=TENSOR2##_start, TENSOR1##_i = TENSOR1##_start; ((count < line_seg_len) && (TENSOR2##_i < TENSOR2##_size) && (TENSOR1##_i < TENSOR1##_size)); ++TENSOR2##_i, ++TENSOR1##_i, ++count){ \
+          CODE                                                                                               \
+          TENSOR2##_data += TENSOR2##_stride;                                                                \
+          TENSOR1##_data += TENSOR1##_stride;                                                                \
+        }                                                                                                    \
+        if (count < line_seg_len){                                                                           \
```

Comment:
A comment here mentioning that the iteration index usually decreases once reaches this part might be helpful as well.